datafusion/optimizer/src/push_down_filter.rs

```diff
@@ -955,6 +965,36 @@ impl PushDownFilter {
     }
 }
 
+/// Convert cross join to join by pushing down filter predicate to the join condition
+fn convert_cross_join_to_inner_join(cross_join: CrossJoin) -> Result<Join> {
+    let CrossJoin { left, right, .. } = cross_join;
+    let join_schema = build_join_schema(left.schema(), right.schema(), &JoinType::Inner)?;
+    // predicate is given
+    Ok(Join {
+        left,
+        right,
+        join_type: JoinType::Inner,
+        join_constraint: JoinConstraint::On,
+        on: vec![],
+        filter: None,
+        schema: DFSchemaRef::new(join_schema),
+        null_equals_null: true,
+    })
+}
+
+/// Converts the inner join with empty equality predicate and empty filter condition to the cross join
+fn convert_to_cross_join_if_beneficial(plan: LogicalPlan) -> Result<LogicalPlan> {
+    if let LogicalPlan::Join(join) = &plan {
+        // Can be converted back to cross join
+        if join.on.is_empty() && join.filter.is_none() {
+            return LogicalPlanBuilder::from(join.left.as_ref().clone())
+                .cross_join(join.right.as_ref().clone())?
+                .build();
+        }
+    }
+    Ok(plan)
+}
```

Comment:
Hmm, after `push_down_all_join`, if there are predicates that cannot be join condition, they will be into a `Filter` on top of join node. In the case, this function will skip to convert the under join to cross join?

Reply:
To be more precise filter can be pushed down below the join completely. In this case we may end up with joins empty equality predicate and empty filter condition.
As an example

```
Filter(l.a>l.b AND r.a>r.b)
--Join (on=[], filter=None)
----LeftTable(a,  b)
----RightTable(a, b)
```
will be converted to the plan below after `push_down_all_join`
```
Join (on=[], filter=None)
--Filter(l.a>l.b)
----LeftTable(a,  b)
--Filter(r.a>r.b)
----RightTable(a, b)
```
this util ensures that plan above is converted to the plan below
```
CrossJoin
--Filter(l.a>l.b)
----LeftTable(a,  b)
--Filter(r.a>r.b)
----RightTable(a, b)
```
However, if the original plan were
```
Filter(l.a>r.b)
--Join (on=[], filter=None)
----LeftTable(a,  b)
----RightTable(a, b)
```
after `push_down_all_join` we will end up with following plan
```
Join (on=[], filter=Some(l.a>r.b))
--LeftTable(a,  b)
--RightTable(a, b)
```
in this case join top cannot be converted to the cross join.

Reply:
I mean if after `push_down_all_join` there are predicates remaining in Filter, but Join has empty `on` and empty `filter`, because you match the plan by `if let LogicalPlan::Join(join) = &plan`, the Join won't be converted to a Cross Join even looks like it should be (based on the logic here).

Reply:
As far as I know, after `after push_down_all_join` predicates remaining in Filter will be pushed into the join filter. Hence, we are sure that top operator is `Join`.

Reply:
This is how `push_down_all_join` handles remaining predicates in Filter. Does 

```rust
if keep_predicates.is_empty() {
  Ok(plan)
} else {
  // wrap the join on the filter whose predicates must be kept
  match conjunction(keep_predicates) {
    Some(predicate) => Ok(LogicalPlan::Filter(Filter::try_new(
      predicate,
      Arc::new(plan),
    )?)),
    None => Ok(plan),
  }
  }
```

And after it, `convert_to_cross_join_if_beneficial` is called. Where does the Filter push down? Am I missing it?

```
let plan = push_down_all_join(
  predicates,
  vec![],
  &join_plan,
  left,
  right,
  vec![],
  true,
  )?;
convert_to_cross_join_if_beneficial(plan)?
```

Reply:
And, the reason that the predicate is kept in Filter in `push_down_all_join` is because it cannot be pushed into Join filter or pushed down through Join. How does it push into/push down later?

Reply:
I think you are right. We will discuss and file a follow-on PR today to fix.